src/mlpack/methods/linear_svm/linear_svm_main.cpp

```diff
@@ -0,0 +1,449 @@
+      const double stepSize = CLI::GetParam<double>("step_size");
+      const bool shuffle = !CLI::HasParam("shuffle");
+      const size_t maxIt = epochs * trainingSet.n_cols;
+
+      ens::ConstantStep decayPolicy(stepSize);
+
+      #ifdef HAS_OPENMP
+      size_t threads = omp_get_max_threads();
+      #else
+      size_t threads = 1;
+      Log::Warn << "Using parallel SGD, but OpenMP support is "
+                << "not available!" << endl;
+      #endif
+
+      ens::ParallelSGD<ens::ConstantStep> psgdOpt(maxIt, std::ceil(
+        (float) trainingSet.n_cols / threads), tolerance, shuffle,
+        decayPolicy);
+
+      Log::Info << "Training model with ParallelSGD optimizer." << endl;
+
+      // This will train the model.
+      model->svm.Train(trainingSet, labels, numClasses, psgdOpt);
+    }
+  }
+  if (CLI::HasParam("test"))
+  {
+    if (!CLI::HasParam("training"))
+    {
+      numClasses = model->svm.NumClasses();
+    }
+    // Get the test dataset, and get predictions.
+    testSet = std::move(CLI::GetParam<arma::mat>("test"));
+    arma::Row<size_t> predictions;
+    size_t trainingDimensionality;
+
+    // Set the dimensionality according to fitintercept.
+    if (intercept)
+      trainingDimensionality = model->svm.Parameters().n_rows - 1;
+    else
+      trainingDimensionality = model->svm.Parameters().n_rows;
+
+    // Checking the dimensionality of the test data.
+    if (testSet.n_rows != trainingDimensionality)
+    {
+      Log::Fatal << "Test data dimensionality (" << testSet.n_rows << ") must "
+          << "be the same as the dimensionality of the training data ("
+          << trainingDimensionality << ")!" << endl;
+    }
+
+    // Save class probabilities, if desired.
+    if (CLI::HasParam("probabilities"))
+    {
+      Log::Info << "Calculating class probabilities of points in "
+          << CLI::GetPrintableParam<arma::mat>("test") << "." << endl;
+      arma::mat probabilities;
+      model->svm.Classify(testSet, probabilities);
+      CLI::GetParam<arma::mat>("probabilities") = std::move(probabilities);
+    }
+
+    model->svm.Classify(testSet, predictedLabels);
+    data::RevertLabels(predictedLabels, model->mappings, predictions);
+
+    // Calculate accuracy, if desired.
+    if (CLI::HasParam("test_labels"))
+    {
+      arma::Row<size_t> testLabels;
+      arma::Row<size_t> testRawLabels =
+          std::move(CLI::GetParam<arma::Row<size_t>>("test_labels"));
+
+      data::NormalizeLabels(testRawLabels, testLabels, model->mappings);
+
+      if (testSet.n_cols != testLabels.n_elem)
+      {
+        Log::Fatal << "Test data given with " << PRINT_PARAM_STRING("test")
+            << " has " << testSet.n_cols << " points, but labels in "
+            << PRINT_PARAM_STRING("test_labels") << " have "
+            << testLabels.n_elem << " labels!" << endl;
+      }
+
+      numClasses = CLI::GetParam<int>("num_classes") == 0 ?
+        model->mappings.n_elem : CLI::GetParam<int>("num_classes");
```

Comment:
Oops, this line should be doubly indented.  I can handle it during merge unless you get to it before me. :)

Reply:
Job done :)